crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -355,6 +367,14 @@ pub trait ReleaseStateQueryData: QueryData {
     fn release_state<'w>(item: Self::Item<'w, '_>) -> Self::Item<'w, 'static>;
 }
 
+/// A marker trait to indicate that the query data filters at an archetype level.
+///
+/// This is needed to implement [`ExactSizeIterator`] for
+/// [`QueryIter`](crate::query::QueryIter) that contains archetype-level filters.
+///
+/// The trait must only be implemented for query data where its corresponding [`QueryData::IS_ARCHETYPAL`] is [`prim@true`].
+pub trait ArchetypeQueryData: QueryData {}
```

Comment:
This would not be needed if we could use nightly and `const_generics`, right?

Reply:
> This would not be needed if we could use nightly and `const_generics`, right?

Oh, I don't know!  ... Yup, it looks like `#![feature(associated_const_equality)]` would let us get rid of this, along with `ArchetypeFilter` and `ReadOnlyQueryData`!  

I think there's also [a proposal for `#![feature(downcast_trait)]`](https://redirect.github.com/rust-lang/rust/issues/144361) that would go the other direction and let us remove the associated constants in favor of the traits.